### DISCUSSION TITLE
docker container: how to add clients and users?

### DISCUSSION THREAD

**Discussion by candlerb** [dexidp/dex#2129] [Q&A]
I have possibly missed something obvious...

The Dockerfile uses [config.docker.yaml](https://github.com/dexidp/dex/blob/master/config.docker.yaml) which brings up Dex with a sqlite database, without any static clients or users.  But it also doesn't enable the gRPC API.  In that case, what's the expected way of adding clients, connectors and/or PasswordDB entries?

I guess I can replace the config file - but that seems to defeat the purpose of the carefully-templated `config.docker.yaml` that is supplied.  Or I could do direct SQL updates on the sqlite3 file?!

Thanks,

Brian.

**Comment by nabokihms** [comment]
Hello @candlerb. You can change the template in the docker container as you want. The config.docker.yaml is more for demonstration purposes than for real-world experience.

**Reply by candlerb** [reply]
Ah OK, I was just wondering if I was missing something obvious.

Clearly some work has gone into inserting suitable environment variables into the template, so it's surprising that that's not sufficient to make it perform any useful function :-)

**Reply by nabokihms** [reply]
Development and maintaining a template that covers all the cases is not an easy thing. We are trying to cut corners.

**Reply by sagikazarmark** [reply]
The idea behind the default template is that it's a minimalistic example, just enough to start Dex. We can't possibly include everything in there. :)

**Reply by candlerb** [reply]
In that case, maybe it's worth putting a comment in there saying "this config will need to be extended before you can use Dex", or words to that effect.

Alternatively: the config could enable the gRPC listener. Then it *would* be possible to use Dex with it.

**Reply by sagikazarmark** [reply]
Like this: https://github.com/dexidp/dex/blob/master/config.docker.yaml#L1 ? :)

Dex is certainly usable without gRPC and in most of the cases it's not necessary at all.

**Comment by candlerb** [comment]
> Like this: https://github.com/dexidp/dex/blob/master/config.docker.yaml#L1 ? :)

Fair enough :-)

Certainly Dex is usable without gRPC: but it's not usable with that config.docker.yaml as-is.  When it said "for development purposes", I thought it meant you could run a test suite against it using that config file.  But since you can't add any clients, any users, or any connectors, there's not much you can do AFAICS.

It has env var features to enable the mock connector - but not any clients.  It has env var features to enable passwordDB - but not any users/passwords.

**Reply by sagikazarmark** [reply]
Well, if you ask me, there are too many options in that example file.

It's only purpose is to let Dex launch (ie. it needs at least one connector or the password DB, otherwise the process fails).

You should always write your own.

**Reply by candlerb** [reply]
Please can you open a separate ticket for your issue.

Since you have shown a config file, I am guessing that you've replaced the default config file that comes with the docker container with your own, and therefore that's a different problem.

Note: it's helpful if you can you put three backticks (` ``` `) on a line of their own, before and after the pasted config, so that it formats properly.

**Comment by iaean** [comment]
We are deriving from `dexidp/dex` and using the common `ENTRYPOINT` pattern to prepare configuration for production.

```docker
FROM dexidp/dex:v2.31.1
# ...
COPY --chown=root:root docker-entrypoint.sh /entrypoint.sh
# ...
ENTRYPOINT ["/entrypoint.sh"]
CMD ["dex", "serve", "config.yaml"]
```
